### DISCUSSION TITLE
How to disable terminal colors in the vitest snashots?

### DISCUSSION THREAD

**Discussion by ghiscoding** [vitest-dev/vitest#6587] [Q&A]
I'm having unit tests issues with my snapshots after switching from `chalk` to `picocolors` in Lerna-Lite via this [PR](https://github.com/lerna-lite/lerna-lite/pull/940). I saw this Vitest issue #841 which says to use `NO_COLOR` which seems to help in the CI and makes my Lerna-Lite [PR](https://github.com/lerna-lite/lerna-lite/pull/940) pass but I can't figure how to disable it on my local Windows computer.

I tried setting this in my `.env` file but that doesn't work
```ts
NO_COLOR=true
FORCE_COLOR=0
```
I also tried the Vitest `--no-color` CLI option but that doesn't work either. I also have the problem while running VSCode debugger as well. It's great that it now works in the CI but I really I could get my tests to pass locally as well.

The code below works in the CI, but how can I do the same locally?

```sh
- name: Run Vitest unit tests
  env:
    NO_COLOR: true
  run: pnpm test:coverage
```

**Comment by ghiscoding** [comment]
ahh wait this seems to work by using `cross-env` package

```s
"scripts": {
  "test": "cross-env-shell NO_COLOR=true vitest --watch --config ./vitest/vitest.config.ts",
}
```
is that the best way to do it on a Windows and to make it cross-compatible?

and this seems to work for the VSCode debugger

```ts
{
  "name": "Vitest - Debug Current Test File",
  "type": "node",
  "env": {
    "NO_COLOR": "true"
  },
  "request": "launch",
  "autoAttachChildProcesses": true,
  "skipFiles": ["<node_internals>/**", "**/node_modules/**"],
  "program": "${workspaceRoot}/node_modules/vitest/vitest.mjs",
  "args": ["run", "${relativeFile}", "--pool", "forks", "--no-watch", "--config", "./vitest/vitest.config.ts"],
  "smartStep": true,
  "console": "integratedTerminal"
},
```

**Comment by AriPerkkio** [comment]
The `NO_COLOR` should work. But if you are testing terminal output, I would recommend to use [`strip-ansi`](https://www.npmjs.com/package/strip-ansi) in the snapshots. It will get rid of colors and other ANSI escape codes. 

Something like this should work nicely:

```ts
import { expect } from 'vitest';
import stripAnsi from 'strip-ansi'

expect.addSnapshotSerializer({
  serialize: (val) => stripAnsi(val),
  test: () => true,
});
```

**Reply by ghiscoding** [reply]
ahh no I wasn't testing the color in the snapshot, I simply wanted to replace chalk and a few tests failing with extra characters because the colors were kept which caused snapshot diffs. Chalk had a `level` option that could be set to 0 for disabling the colors which is what was used earlier. So, anyway I got it working with cross-env and NO_COLOR in the npm script, so I'll use that unless I find something better. Thanks

**Comment by mrazauskas** [comment]
Since Node.js v16.11 the built in `util.stripVTControlCharacters()` method could be used instead of `strip-ansi`. It does the same thing. Reference: https://nodejs.org/docs/latest/api/util.html#utilstripvtcontrolcharactersstr

**Reply by ghiscoding** [reply]
@mrazauskas  That's nice, I wonder where would that be used? I guess it could be interesting to see implemented directly into Vitest to better disable colors, is that what you had in mind?

**Reply by mrazauskas** [reply]
It can be used instead of `strip-ansi`. This package has nearly 200 million downloads a week and more than 9000 dependants. Hope this answers your first question.

---

Sorry, but I cannot answer your second question seriously. Somehow it feels like Vitest already is a JavaScript testing tool with the largest API surface. And maintainers continue to think about adding more: https://github.com/vitest-dev/vitest/issues/5857#issuecomment-2376718017

> **Vitest. More is more!**

I would like to propose a new tag line. Because “Next Generation” doesn't really say anything. “A Vite-native testing framework“ is plain wrong, because neither benchmarks nor type checks are anyhow Vite-native.

So what it is?

Perhaps: “Vitest. More is more! A JavaScript testing tool with the largest API surface ever. (And we are eager for more!)”

**Comment by kmturley** [comment]
I am using https://www.npmjs.com/package/cli-table3 library to format my cli output.

Here is an example of the snapshot produced:
<img width="779" alt="Screenshot 2025-03-10 at 4 21 30 PM" src="https://github.com/user-attachments/assets/2c11028c-668c-40ca-8f09-11ddd12d0c1a" />

Snapshot matches on MacOS locally, but fails when run on Github Actions (even on MacOS runner)
<img width="837" alt="Screenshot 2025-03-10 at 4 10 12 PM" src="https://github.com/user-attachments/assets/c69c4065-04ec-4e1d-bc9a-bd3ea9b19982" />

If I include the recommended fix:
```
import { stripVTControlCharacters } from 'util';

expect.addSnapshotSerializer({
  serialize: (val) => stripVTControlCharacters(val),
  test: () => true,
});
```

Then delete/recreate the snapshot file, it comes out as:
<img width="803" alt="Screenshot 2025-03-10 at 4 32 50 PM" src="https://github.com/user-attachments/assets/d3a42e55-34ec-4331-a819-568def937f70" />

Which works nicely!
